pilot/pkg/networking/core/v1alpha3/listener.go

```diff
@@ -580,45 +640,37 @@ func (configgen *ConfigGeneratorImpl) buildSidecarInboundListenerForPortOrUDS(no
 		switch pluginParams.ListenerProtocol {
 		case plugin.ListenerProtocolHTTP:
 			filterChainMatch = chain.FilterChainMatch
+			// This is the filter chain used by permissive mTLS. Append applicationProtocolsOverTLS as the client side will
+			// override the ALPN with applicationProtocolsOverTLS.
+			if filterChainMatch != nil && len(filterChainMatch.ApplicationProtocols) > 0 {
+				filterChainMatch.ApplicationProtocols = append(filterChainMatch.ApplicationProtocols, applicationProtocolsOverTLS...)
+			}
 			httpOpts = configgen.buildSidecarInboundHTTPListenerOptsForPortOrUDS(node, pluginParams)
 
 		case plugin.ListenerProtocolTCP:
 			filterChainMatch = chain.FilterChainMatch
 			tcpNetworkFilters = buildInboundNetworkFilters(pluginParams.Env, pluginParams.Node, pluginParams.ServiceInstance)
 
 		case plugin.ListenerProtocolAuto:
-			// TODO(crazyxy) avoid bypassing authN using TCP
-			// Build filter chain options for listener configured with protocol sniffing
-			// Double the number of filter chains. Half of filter chains are used as http filter chain and half of them are used as tcp proxy
-			// id in [0, len(allChains)/2) are configured as http filter chain, [(len(allChains)/2, len(allChains)) are configured as tcp proxy
-			// If mTLS is enabled, there are four filter chains. The filter chain match should be
-			//  FCM 1: ALPN [istio, http/1.0, http/1.1, h2] Transport protocol: tls
-			//  FCM 2: ALPN [http/1.0, http/1.1, h2] Transport protocol: N/A
-			//  FCM 3: ALPN [istio] Transport protocol: N/A
-			//  FCM 4: ALPN [] Transport protocol: N/A
-			// If mTLS is disabled, there are two filter chains. The filter chaim match should be
-			//  FCM 1: ALPN [http/1.0, http/1.1, h2]
-			//  FCM 2: ALPN []
-			if id < len(allChains)/2 {
-				httpOpts = configgen.buildSidecarInboundHTTPListenerOptsForPortOrUDS(node, pluginParams)
-
-				fcm := listener.FilterChainMatch{}
-				if chain.FilterChainMatch != nil {
-					fcm = *chain.FilterChainMatch
-				}
+			if filterChainMatchOption == nil || len(filterChainMatchOption) <= id {
```

Comment:
what is this check doing? this match based on array index is not that explanatory